subsys/net/lib/sockets/sockets_tls.c

```diff
@@ -5,18 +5,176 @@
+
+/** TLS context information. */
+struct tls_context {
+	/** Information whether TLS context is used. */
+	bool is_used;
+
+	/** Secure protocol version running on TLS context. */
+	enum net_ip_protocol_secure tls_version;
+
+#if defined(CONFIG_MBEDTLS)
+	/** mbedTLS context. */
+	mbedtls_ssl_context ssl;
+
+	/** mbedTLS configuration. */
+	mbedtls_ssl_config config;
+#endif /* CONFIG_MBEDTLS */
+};
+
+/* A global pool of TLS contexts. */
+static struct tls_context tls_contexts[CONFIG_NET_SOCKETS_TLS_MAX_CONTEXTS];
+
+/* A mutex for protecting TLS context allocation. */
+static struct k_mutex context_lock;
+
+/* Initialize TLS internals. */
+static int tls_init(struct device *unused)
+{
+	ARG_UNUSED(unused);
+
+	memset(tls_contexts, 0, sizeof(tls_contexts));
+
+	k_mutex_init(&context_lock);
+
+	return 0;
+}
+
+SYS_INIT(tls_init, APPLICATION, CONFIG_KERNEL_INIT_PRIORITY_DEFAULT);
+
+/* Allocate TLS context. */
+static struct tls_context *tls_alloc(void)
+{
+	int i;
+	struct tls_context *tls = NULL;
+
+	k_mutex_lock(&context_lock, K_FOREVER);
+
+	for (i = 0; i < ARRAY_SIZE(tls_contexts); i++) {
+		if (!tls_contexts[i].is_used) {
+			tls = &tls_contexts[i];
+			memset(tls, 0, sizeof(*tls));
+			tls->is_used = true;
+
+			NET_DBG("Allocated TLS context, %p", tls);
+			break;
+		}
+	}
+
+	k_mutex_unlock(&context_lock);
+
+	if (tls) {
+		mbedtls_ssl_init(&tls->ssl);
+		mbedtls_ssl_config_init(&tls->config);
+	} else {
+		NET_WARN("Failed to allocate TLS context");
+	}
+
+	return tls;
+}
+
+/* Release TLS context. */
+static int tls_release(struct tls_context *tls)
+{
+	if (!PART_OF_ARRAY(tls_contexts, tls)) {
+		NET_ERR("Invalid TLS context");
+		return -EINVAL;
+	}
+
+	if (!tls->is_used) {
+		NET_ERR("Deallocating unused TLS context");
+		return -EACCES;
+	}
+
+	mbedtls_ssl_config_free(&tls->config);
+	mbedtls_ssl_free(&tls->ssl);
+
+	k_mutex_lock(&context_lock, K_FOREVER);
+
+	tls->is_used = false;
+
+	k_mutex_unlock(&context_lock);
+
+	return 0;
+}
 
 int ztls_socket(int family, int type, int proto)
 {
-	return zsock_socket(family, type, proto);
+	enum net_ip_protocol_secure tls_proto = 0;
+	int sock, ret, err;
+
+	if (proto  >= IPPROTO_TLS_1_0 && proto <= IPPROTO_TLS_1_2) {
+		tls_proto = proto;
+		proto = (type == SOCK_STREAM) ? IPPROTO_TCP : IPPROTO_UDP;
+	}
+
+	sock = zsock_socket(family, type, proto);
+	if (sock < 0) {
+		/* errno will be propagated */
+		return -1;
+	}
+
+	if (tls_proto != 0) {
+		/* If TLS protocol is used, allocate TLS context */
+		struct net_context *context = INT_TO_POINTER(sock);
+
+		context->tls = tls_alloc();
+
+		if (!context->tls) {
+			ret = -ENOMEM;
+			goto error;
+		}
+
+		context->tls->tls_version = tls_proto;
+	}
+
+	return sock;
+
+error:
+	err = zsock_close(sock);
+	__ASSERT(err == 0, "Socket close failed");
+
+	errno = -ret;
+	return -1;
 }
 
 int ztls_close(int sock)
 {
+	struct net_context *context = INT_TO_POINTER(sock);
+	int ret;
+
+	if (context->tls) {
+		ret = tls_release(context->tls);
```

Comment:
So, given EBADF comments above, the way to handle errors might be more involved. It should finalize both TLS context and socket, but if any returned error, it should return it too (with errno set appropriately). If both error out, it's up to you to choose which prevails ;-).